synapse/lib/cell.py

```diff
@@ -2093,6 +2091,57 @@ async def reqGateKeys(self, gatekeys):
         for useriden, perm, gateiden in gatekeys:
             (await self.auth.reqUser(useriden)).confirm(perm, gateiden=gateiden)
 
+    async def feedBeholder(self, name, info, gates=None, perms=None):
+        '''
+        Feed a named event onto the beholder message bus that will sent to any listeners.
+
+        Args:
+            info (dict): An information dictionary to be sent to any consumers.
+            gates (list): List of gate idens, whose details will be added to the outbound message(s).
+            perms (list): List of permission names, whose details will be added to the outbound message(s).
+
+        Returns:
+            None
+        '''
+        kwargs = {
+            'event': name,
+            'offset': await self.nexsroot.index(),
+            'info': copy.deepcopy(info),
+        }
+
+        if gates:
+            g = []
+            for gate in gates:
+                authgate = await self.getAuthGate(gate)
+                if authgate is not None:
+                    g.append(authgate)
+            kwargs['gates'] = g
+
+        if perms:
+            p = []
+            for perm in perms:
+                permdef = await self.getPermDef(perm)
+                if permdef is not None:
+                    p.append(permdef)
+            kwargs['perms'] = p
+
+        await self.fire('cell:beholder', **kwargs)
+
+    @contextlib.asynccontextmanager
+    async def beholder(self):
+        async with await s_queue.Window.anit(maxsize=10000) as wind:
+            async def onEvent(mesg):
+                await wind.put(mesg[1])
+
+            self.on('cell:beholder', onEvent, base=wind)
```

Comment:
we have an `onWith` handler which is a context manager that will automatically remove the `cell:beholder` handler. we should use that guy here.